fdbserver/ptxn/test/TestTLogServer.actor.cpp

```diff
@@ -845,3 +845,164 @@ TEST_CASE("/fdbserver/ptxn/test/read_persisted_disk_on_tlog") {
+		          << "\n";
+	}
+
+	// replace fake TLogInterface with recruited interface
+	std::vector<Future<ptxn::TLogInterface_PassivelyPull>> interfaceFutures(pContext->numTLogs);
+	for (i = 0; i < pContext->numTLogs; i++) {
+		interfaceFutures[i] = tLogInitializations[i].reply.getFuture();
+	}
+	std::vector<ptxn::TLogInterface_PassivelyPull> interfaces = wait(getAll(interfaceFutures));
+	for (i = 0; i < pContext->numTLogs; i++) {
+		*(pContext->tLogInterfaces[i]) = interfaces[i];
+	}
+
+	for (auto& [tLogGroupID, tLogGroupLeader] : pContext->tLogGroupLeaders) {
+		tLogGroupLeader = pContext->tLogInterfaces[pContext->groupToLeaderId[tLogGroupID]];
+	}
+
+	state IKeyValueStore* d = ds[pContext->storageTeamIDTLogGroupIDMapper[storageTeamID]];
+
+	state std::pair<std::vector<Standalone<StringRef>>, std::vector<Version>> res =
+	    wait(commitInjectReturnVersions(pContext, storageTeamID, pContext->numCommits));
+	wait(verifyPeek(pContext, storageTeamID, pContext->numCommits));
+
+	// wait here so that actors who update persistentData can do their job.
+	wait(delay(1.5));
+
+	// Start to recover, put the same tlog groups in the requests as initial assignment
+	state ptxn::TLogGroupID targetGroup = pContext->storageTeamIDTLogGroupIDMapper[storageTeamID];
+	state std::unordered_map<ptxn::TLogGroupID, std::pair<IKeyValueStore*, IDiskQueue*>> dqs;
+	state int writtenTLogID = pContext->groupToLeaderId[pContext->storageTeamIDTLogGroupIDMapper[storageTeamID]];
+	PromiseStream<ptxn::InitializePtxnTLogRequest> initializeTLogRecover;
+
+	state IDiskQueue::location previousNextPushLocation = qs[targetGroup]->getNextPushLocation();
+
+	tLogInitializations.emplace_back();
+	tLogInitializations.back().isPrimary = true;
+	tLogInitializations.back().storeType = KeyValueStoreType::MEMORY;
+	tLogInitializations.back().tlogGroups = pContext->groupsPerTLog[writtenTLogID];
+	// nede to set recruitementId to avoid caching
+	tLogInitializations.back().recruitmentID = ptxn::test::randomUID();
+	ptxn::InitializePtxnTLogRequest req = tLogInitializations.back();
+	const StringRef prefix = req.logVersion > TLogVersion::V2 ? fileVersionedLogDataPrefix : fileLogDataPrefix;
+
+	for (ptxn::TLogGroup& tlogGroup : pContext->groupsPerTLog[writtenTLogID]) {
+		std::string filename = filenameFromId(req.storeType, folder, prefix.toString() + "test", tlogGroup.logGroupId);
+		IKeyValueStore* data = openKVStore(req.storeType, filename, tlogGroup.logGroupId, 500e6);
+		IDiskQueue* queue = openDiskQueue(joinPath(folder, diskQueueFilePrefix + tlogGroup.logGroupId.toString() + "-"),
+		                                  "fdq",
+		                                  tlogGroup.logGroupId,
+		                                  DiskQueueVersion::V1);
+		dqs[tlogGroup.logGroupId] = std::make_pair(data, queue);
+	}
+
+	// cancel all actors, but disk files would not be erase so that we can recover from it.
```

Comment:
Since the `ptxn::tLog` actor is cancelled here, the actual recovery happened after this loop. Probably worth mentioning the tlog is shutdown here.